lib/kredis/types/ordered_set.rb

```diff
@@ -0,0 +1,71 @@
+class Kredis::Types::OrderedSet < Kredis::Types::Proxying
+  proxying :multi, :zrange, :zrem, :zadd, :zremrangebyrank, :zcard, :exists?, :del
+
+  attr_accessor :typed
+  attr_reader :limit
+
+  def elements
+    strings_to_types(zrange(0, -1) || [], typed)
+  end
+  alias to_a elements
+
+  def remove(*elements)
+    zrem(types_to_strings(elements, typed))
+  end
+
+  def prepend(elements)
+    insert(elements, prepending: true)
+  end
+
+  def append(elements)
+    insert(elements)
+  end
+  alias << append
+
+  def limit=(limit)
+    raise "Limit must be greater than 0" if limit && limit <= 0
+
+    @limit = limit
+  end
+
+  private
+    def insert(elements, prepending: false)
+      elements = Array(elements)
+      return if elements.empty?
+
+      elements_with_scores = types_to_strings(elements, typed).map.with_index do |element, index|
+        score = generate_base_score(negative: prepending) + index / 100000
+
+        [ score , element ]
+      end
+
+      multi do |pipeline|
+        pipeline.zadd(elements_with_scores)
+        trim(from_beginning: prepending, pipeline: pipeline)
+      end
+    end
+
+    def generate_base_score(negative:)
+      current_time = process_start_time + process_uptime
+
+      negative ? -current_time : current_time
+    end
```

Comment:
Clever approach 👍!  

Just a note for your consideration: 

What would happen in case of skewed Redis clocks / concurrent writes? Is possible to face the scenario where appending 2 arrays would lead to mixing their elements?

```ruby
# In process A
ordered_set.append([1, 2, 3])
# In process B
ordered_set.append([4, 5, 6])
ordered_set.elements
# Is this possible?
# [1, 2, 4, 3, 5, 6]
```

Maybe a less clever approach that relies getting the max/min (with an optimistic lock) would be more resilient?

```
WATCH zset
# get max
element = ZRANGE zset -1 -1
# compute scores max +1 for each element
MULTI
# add elements
ZADD zset elemens
# trim
EXEC
```